Doc/library/typing.rst

```diff
@@ -1230,6 +1232,103 @@ These are not used in annotations. They are building blocks for creating generic
     ``covariant=True`` or ``contravariant=True``.  See :pep:`484` for more
     details.  By default, type variables are invariant.
 
+.. class:: TypeVarTuple
+
+    :class:`Type variable <TypeVar>` tuple.
+
+    A vanilla type variable enables parameterization with a single type. A type
+    variable tuple, in contrast, allows parameterization with an
+    *arbitrary* number of types by acting like an *arbitrary* number of type
+    variables wrapped in a tuple. For example::
+
+        Shape = TypeVarTuple('Shape')
+
+        class Array(Generic[*Shape]): ...
+
+        array_1d: Array[int] = Array()       # 1 type parameter: fine
+        array_2d: Array[int, int] = Array()  # 2 type parameters: also fine
+        array_0d: Array[()] = Array()        # 0 type parameters: also fine
+
+    Note the use of the unpacking operator ``*`` in ``Generic[*Shape]``.
+    Conceptually, you can think of ``Shape`` as a tuple of type variables
+    ``(T1, T2, ...)``. ``Generic[*Shape]`` would then become
+    ``Generic[*(T1, T2, ...)]``. The ``*`` brings the type variables out of the
+    tuple and into the square brackets directly.
+
+    (In older versions of Python, where ``*`` was less flexible, you might see
+    this written using :data:`Unpack <Unpack>` instead, as
+    ``Unpack[Shape]``. This means the same thing as ``*Shape`` - ``Unpack`` and
+    ``*`` can be used interchangeably in the context of types.)
```

Comment:
It was my idea and I still feel it's a good idea to at least bring up `Unpack` as that is likely what people reading these docs are going to write for a few releases at least.

(What's wrong with parentheses?)

Reply:
Parentheses are great but should be used with caution (they disrupt the flow of the sentence a little). They can also come across as a little informal sometimes (it's almost like an aside to the reader).

But more specifically in this instance -- a whole paragraph in parentheses, to me, seems a bit odd for technical documentation. Parentheses should be used for inessential, skippable sections. But if this section is inessential, then it probably shouldn't be included in the docs at all -- and if it's essential, then it probably shouldn't be in parentheses!

Reply:
Well, in a sense this section *is* skippable if you're only using Python 3.11. Removing the parentheses to me feels like it would emphasize Unpack too much. Using a markup like "Note" just draws *more* attention to it. And omitting the paragraph feels doing users a disservice, because for many this *will* come up.

Maybe we'll just have to agree to disagree on how (in)formal these docs are supposed to be.

Reply:
I think my main issue is that this paragraph duplicates the section on `Unpack` lower down. I don't mind referencing `Unpack` and linking to the dedicated documentation on `Unpack`, but I think this (as a sentence tacked onto the previous paragraph, rather than as its own paragraph) would be sufficient:

```suggestion
    (Note that in older versions of Python, you might see this written using
    :data:`Unpack <Unpack>` instead, as  ``Unpack[Shape]``.)
```

Reply:
Mm, fair point about the duplication with Unpack. Have replaced this with your suggestion.